Add examples to Key value page (#3105)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/processors/key-value.md

## Overview

You can use the `key_value` processor to parse the specified field into key-value pairs. You can customize the `key_value` processor to parse field information with the following options.

> | Option | Required | Type | Default value | Description | Example |

Reviewer: This table is very wide and is rendered with a scroll bar even for fairly large screen sizes. To make it  narrower, consider removing the "type" column and just saying in the description "All options are of type string" or something like that. Also, we can remove the "required" column and the "default value" column and just add a statement to the description on whether the option is required. For example, the description for the first line of the table will read "The key in the event that is parsed. Optional. Default is `message`".

Author: Originally, the "required" and "default value" information was contained in the description field and was confusing to find that information in the "Description" column. Removing the "Type" column simplifies the appearance significantly.

| :--- | :--- | :--- | :--- | :--- |

> | source | No | String | `message` | The key in the event that is parsed. | `source` is `"message1"`. `{"message1": {"key1=value1"}, "message2": {"key2=value2"}}` parses into `{"message1": {"key1=value1"}, "message2": {"key2=value2"}, "parsed_message": {"key1": "value1"}}`. |

Reviewer: ```suggestion
| source | No | String | `message` | The message field to be parsed.  | `source` is `"message1"`. `{"message1": {"key1=value1"}, "message2": {"key2=value2"}}` parses into `{"message1": {"key1=value1"}, "message2": {"key2=value2"}, "parsed_message": {"key1": "value1"}}`. |
```

Reviewer: When you remove the "type", "required", and "default value" columns, see if the examples are formatted nicely in one line. If they're not, I suggest breaking them out into a section below the table. In any case, make the examples complete sentences.

> | destination | No | String | `parsed_message` | The destination key for the parsed source output. Outputting the parsed source overwrites the value of the key if it already exists. | `destination` is `"parsed_data"`. `{"message": {"key1=value1"}}` parses into `{"message": {"key1=value1"}, "parsed_data": {"key1": "value1"}}`. |

Reviewer: ```suggestion
| destination | No | String | `parsed_message` | The destination field for the parsed source. The parsed source overwrites preexisting data for that key. | `destination` is `"parsed_data"`. `{"message": {"key1=value1"}}` parses into `{"message": {"key1=value1"}, "parsed_data": {"key1": "value1"}}`. |
```

> | field_delimiter_regex | Conditionally | String | N/A | A regular expression (regex) specifying the delimiter between key-value pairs. Special regex characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `field_split_characters`. | `field_delimiter_regex` is `"&\\{2\\}"`. `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |

Reviewer: ```suggestion
| field_delimiter_regex | Conditionally | String | N/A | A regular expression specifying the delimiter that separates key-value pairs. Special characters such as `[` and `]` must be escaped with `\\`. This option cannot be defined at the same time as `field_split_characters`. | `field_delimiter_regex` is `"&\\{2\\}"`. `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |
```

> | field_split_characters | Conditionally | String | `&` | A string of characters to split between key-value pairs. Special regex characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `field_delimiter_regex`. | `field_split_characters` is `"&&"`. `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |

Reviewer: ```suggestion
| field_split_characters | Conditionally | String | `&` | A string of characters specifying the delimiter that separates key-value pairs. Special characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `field_delimiter_regex`. | `field_split_characters` is `"&&"`. `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |
```

> | field_split_characters | A string of characters specifying the delimeter that separates key-value pairs. Special regular expression characters such as `[` and `]` must be escaped with `\\`. This option cannot be defined at the same time as `field_delimiter_regex`. Required conditionally. Default value is `&`. | If `field_split_characters` is `"&&"`, `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |

Reviewer: ```suggestion
| field_split_characters | A string of characters specifying the delimeter that separates key-value pairs. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Cannot be defined at the same time as `field_delimiter_regex`. Optional. Default value is `&`. | If `field_split_characters` is `"&&"`, `{"key1=value1&&key2=value2"}` parses into `{"key1": "value1", "key2": "value2"}`. |
```

> | key_value_delimiter_regex| Conditionally | String | N/A | A regex specifying the delimiter between a key and a value. Special regex characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `value_split_characters`. | `key_value_delimiter_regex` is `"=\\{2\\}"`. `{"key1==value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| key_value_delimiter_regex| Conditionally | String | N/A | A regex specifying the delimiter that separates a key and a value within a key-value pair. Special characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `value_split_characters`. | `key_value_delimiter_regex` is `"=\\{2\\}"`. `{"key1==value1"}` parses into `{"key1": "value1"}`. |
```

> | key_value_delimiter_regex | A regular expression specifying that separates a key and a value within a key-value pair. Special regular expression characters such as `[` and `]` must be escaped with `\\`. This option cannot be defined at the same time as `value_split_characters`. Required conditionally. No default value. | If `key_value_delimiter_regex` is `"=\\{2\\}"`, `{"key1==value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| key_value_delimiter_regex | A regular expression specifying the delimiter that separates the key and value within a key-value pair. Special regular expression characters such as `[` and `]` must be escaped with `\\`. This option cannot be defined at the same time as `value_split_characters`.  Optional. If this option is not defined, `value_split_characters` is used.  | If `key_value_delimiter_regex` is `"=\\{2\\}"`, `{"key1==value1"}` parses into `{"key1": "value1"}`. |
```

> | value_split_characters | Conditionally | String | `=` | A string of characters to split between keys and values. Special regex characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `key_value_delimiter_regex`. | `value_split_characters` is `"=="`. `{"key1==value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| value_split_characters | Conditionally | String | `=` | A string of characters specifying the delimiter that separates a key and value within a key-value pair. Special characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `key_value_delimiter_regex`. | `value_split_characters` is `"=="`. `{"key1==value1"}` parses into `{"key1": "value1"}`. |
```

> | value_split_characters | A string of characters specifying the delimiter that separates a key and value within a key-value pair. Special regular expression characters such as `[` and `]` must be escaped with `\\`. This option cannot be defined at the same time as `key_value_delimiter_regex`. Required conditionally. Default value is `=`. | If `value_split_characters` is `"=="`, `{"key1==value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| value_split_characters | A string of characters specifying the delimiter that separates the key and value within a key-value pair. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Cannot be defined at the same time as `key_value_delimiter_regex`. Optional. Default value is `=`. | If `value_split_characters` is `"=="`, `{"key1==value1"}` parses into `{"key1": "value1"}`. |
```

> | non_match_value | No | String | `null` | When a key-value cannot be successfully split, the key-value is placed in the `key` field, and the specified value is placed in the value field. | `key1value1&key2=value2` parses into `{"key1value1": null, "key2": "value2"}`. |

Reviewer: ```suggestion
| non_match_value | No | String | `null` | When a key-value pair cannot be successfully split, the key-value pair is placed in the `key` field, and the specified value is placed in the value field. | `key1value1&key2=value2` parses into `{"key1value1": null, "key2": "value2"}`. |
```

> | prefix | No | String | Empty string | A prefix given to all keys. | `prefix` is `"custom"`. `{"key1=value1"}` parses into `{"customkey1": "value1"}`. |

Reviewer: ```suggestion
| prefix | No | String | Empty string | A prefix to append before all keys. | `prefix` is `"custom"`. `{"key1=value1"}` parses into `{"customkey1": "value1"}`. |
```

> | non_match_value | When a key-value cannot be successfully split, the key-value pair is placed in the `key` field, and the specified value is placed in the value field. Optional. Default value is `null`. | `key1value1&key2=value2` parses into `{"key1value1": null, "key2": "value2"}`. |

Reviewer: ```suggestion
| non_match_value | When a key-value pair cannot be successfully split, the key-value pair is placed in the `key` field and the specified value is placed in the `value` field. Optional. Default value is `null`. | `key1value1&key2=value2` parses into `{"key1value1": null, "key2": "value2"}`. |
```

> | delete_key_regex | No | String | N/A | A regex used to delete characters from the key. Special regex characters such as `[` and `]` must be escaped using `\\`. | `delete_key_regex` is `"\s"`. `{"key1 =value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| delete_key_regex | No | String | N/A | A regular expression specifying the characters to delete from the key. Special characters such as `[` and `]` must be escaped using `\\`. | `delete_key_regex` is `"\s"`. `{"key1 =value1"}` parses into `{"key1": "value1"}`. |
```

> | prefix | A prefix to append before all keys. Optional. Default value is an empty `string`. | If `prefix` is `"custom"` and `{"key1=value1"}`, it parses into `{"customkey1": "value1"}`.|

Reviewer: ```suggestion
| prefix | A prefix to append before all keys. Optional. Default value is an empty string. | If `prefix` is `"custom"`, `{"key1=value1"}` parses into `{"customkey1": "value1"}`.|
```

> | delete_value_regex | No | String | N/A | A regex used to delete characters from the value. Special regex characters such as `[` and `]` must be escaped using `\\`. | `delete_value_regex` is `"\s"`. `{"key1=value1 "}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| delete_value_regex | No | String | N/A | A regular expression specifying the characters to delete from the value. Special characters such as `[` and `]` must be escaped using `\\`. | `delete_value_regex` is `"\s"`. `{"key1=value1 "}` parses into `{"key1": "value1"}`. |
```

> | delete_key_regex | A regular expression specifying the characters to delete from the key. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Optional. No default value. | If `delete_key_regex` is `"\s"`, `{"key1 =value1"}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| delete_key_regex | A regular expression specifying the characters to delete from the key. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Cannot be an empty string. Optional. No default value. | If `delete_key_regex` is `"\s"`, `{"key1 =value1"}` parses into `{"key1": "value1"}`. |
```

> | delete_value_regex | A regular expression specifying the characters to delete from the value. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Optional. No default value. | If `delete_value_regex` is `"\s"`, `{"key1=value1 "}` parses into `{"key1": "value1"}`. |

Reviewer: ```suggestion
| delete_value_regex | A regular expression specifying the characters to delete from the value. Special regular expression characters such as `[` and `]` must be escaped with `\\`. Cannot be an empty string. Optional. No default value. | If `delete_value_regex` is `"\s"`, `{"key1=value1 "}` parses into `{"key1": "value1"}`. |
```

Reviewer: In my opinion, since there's a lot of text here and the examples are long, it is better to list them in a list as opposed to a table. That way, each point in the description can become a sublist. I much prefer the layout in the repo https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/key-value-processor/README.md. @natebower What do you think?

Author: Visually, it is much easier to parse through these options in a tabular format. I am opting to keep the table format for now.

field_delimiter_regex | Conditionally | String | A regex specifying the delimiter between key/value pairs. Special regex characters such as `[` and `]` must be escaped using `\\`. This option cannot be defined at the same time as `field_split_characters`. |
field_split_characters | Conditionally | String | A string of characters to split between key/value pairs. Special regex characters such as `[` and `]` must be escaped using `\\`. Default value is `&`. This option cannot be defined at the same time as `field_delimiter_regex`. |
key_value_delimiter_regex| Conditionally | String | A regex specifying the delimiter between a key and a value. Special regex characters such as `[` and `]` must be escaped using `\\`. There is no default value. This option cannot be defined at the same time as `value_split_characters`. |
value_split_characters | Conditionally | String | A string of characters to split between keys and values. Special regex characters such as `[` and `]` must be escaped using `\\`. Default value is `=`. This option cannot be defined at the same time as `key_value_delimiter_regex`. |

> non_match_value | No | String | When a key/value cannot be successfully split, the key/value is placed in the `key` field, and the specified value is placed in the value field. Default value is `null`. | `key1value1&key2=value2` parses into `{"key1value1": null, "key2": "value2"}`. |

Author: Include examples section after this table that summarizes examples. Should we include a configuration reference before or after these examples?

Author: Example syntax: pipeline configuration, input event, output event.
Refer to this for formatting: https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/key-value-processor/README.md#basic-usage

Author: @asifsmohammed Are there any examples beyond the three that are provided in this doc?

Author: @asifsmohammed Thanks for taking another look at this, I'll monitor for updates on the examples as discussed today.
